### PR TITLE
Add Alerts on OpsGenie

### PR DESCRIPTION
**Pull request by ptorrezao**
Hi Adrian, thanks for the initiative.

The expansion was the implementation of the GET method for the Alerts endpoint, retrieving 1 message per alert.

The purpose of the expansion is to start including the OpsGenie on my home automation stuff.
Yep, the notifications on the APP and/or the call aren't enough :)

Jokes aside, one of the use cases is turn on of the Lights/turn on the coffee maker when I receive an alert.
All of this without waking up the kid, or annoy too much the wife.
This is helpful for me, it may be helpful for some else.

### CODE REVIEW

**Comment by edreanernst** [comment]
Hi @ptorrezao !

Thanks for contributing! The capability to retrieve alerts is very valuable. I like your use case!

However, it would be better if we didn't create a separate node for that. Instead we should extend the current OpsgenieAlert node to handle that functionality so that the OpsgenieAlert node eventually may handle all aspects of the Opsgenie `/alerts` API endpoint. That way we minimise the number of different nodes to be used.

I will appreciate it if you could adapt your contribution by making the necessary changes to the existing node :
- new getAlerts() method in opsgenie_api.js (which you already have done)
- new case for the `msg.action` switch called `get`
- change opsgenie_alert.html to change `outputs:1`
- document the usage (with an example) in opsgenie_alert.html
- documentation changes to README.md

Feel free to update this Pull Request after making the necessary changes.

Thanks again for your willingness to contribute!

**Comment by ptorrezao** [comment]
Hi @edreanernst,

In fairness I don't agree with having one single Node for all of the Interactions, but yet the project is yours so you're calling the shots. 😃 
I think opsgenie_api.js should cover every interaction(and it is) with the API and each node tackle the specific of the interaction.
Really believe on separation of concerns.
But hey I will refactor to fall into your requirements.

**Comment by edreanernst** [comment]
The original idea was to have one node per endpoint which is also the way I described it in the README. The problem is you can keep separating concerns until you have too many different nodes. It makes sense to group some concerns together. It's a balancing act.

I am warming up to the idea of having separate nodes for sending data to Opsgenie and for receiving data from Opsgenie. In other words, for each API endpoint you will have two nodes - one for reading and one for writing.

We will then have to change the naming at least of the current OpsgenieAlert node to something like OpsgenieAlertsWrite, and then your contributed node to something like OpsgenieAlertsRead. I'll have a proper look at it as soon as I can.

In the meantime, don't refactor yet. I'm also open to more suggestions if you have any.

**Comment by ptorrezao** [comment]
The requirement documentation is still valid thought 🤦‍♂️ 
On the next couple days(or weeks) I will do that at least.

But regarding the "Grand Scheme" I do understand that would be hard to keep track of a API Client on nodered, so we should run away from having a node for Close Alert and another Open Alert. (Would be nice but its unmanageable, on my opinion).
But one Node for GET actions and another for POST/PUT I think its bearable.

**Comment by edreanernst** [comment]
Hi @ptorrezao ,

Hope all is well.

I have made some minor changes to the existing Alert node, basically renaming it to Opsgenie Alert Action. The purpose of the existing node would then be for all alert actions that need to be taken. The new node which you are contributing here would then be called Opsgenie Alert Read, which can be used to get all alerts.

I suggest you pull in the latest master commit. If you could then just name the new node as mentioned above and make the necessary documentation changes, I would then be happy to look at an updated pull request.

Thanks for your effort so far!

**Comment by ptorrezao** [comment]
Hi @edreanernst ,

Since the vision is have only 1 node per action, I refactor the node to accept the action so in the future this node can be complemented (e.g. Get Specific alert)

File: README.md [hunk 1]
```diff
@@ -28,13 +28,16 @@ Alternatively, submit an issue on the [GitHub project page](https://github.com/e
       - [msg.payload](#msgpayload)
       - [msg.identifier](#msgidentifier)
       - [msg.identifierType](#msgidentifiertype)
+    - [Read Action](#alert-reader)
```
Comment (edreanernst) [inline README.md]:
```suggestion
    - [Alert Read](#alert-reader)
```

File: README.md [hunk 2]
```diff
@@ -129,4 +132,93 @@ Please refer to the Opsgenie API documentation for descriptions of each of these
         "note": "Node was here!"
     }
 }
-```
\ No newline at end of file
+```
+---
+
+### Alert Reader
```
Comment (edreanernst) [inline README.md]:
```suggestion
### Alert Read
```

File: README.md [hunk 3]
```diff
@@ -129,4 +132,93 @@ Please refer to the Opsgenie API documentation for descriptions of each of these
...
+
+This node will pull the existing alerts from your OpsGennie Setup according the configuration.
```
Comment (edreanernst) [inline README.md]:
```suggestion
This node will pull the existing alerts from your Opsgenie setup according the configuration.
```

File: README.md [hunk 4]
```diff
@@ -129,4 +132,93 @@ Please refer to the Opsgenie API documentation for descriptions of each of these
...
+
+#### Read msg.action
+This is a string field where an the type of question is can be specified (e.g. alerts)
```
Comment (edreanernst) [inline README.md]:
```suggestion
This is a string field where the type of action can be specified (e.g. list)
```

File: README.md [hunk 5]
```diff
@@ -129,4 +132,93 @@ Please refer to the Opsgenie API documentation for descriptions of each of these
...
+Possible values are:
+* "alerts"
```
Comment (edreanernst) [inline README.md]:
```suggestion
* "list"
```

File: README.md [hunk 6]
```diff
@@ -129,4 +132,93 @@ Please refer to the Opsgenie API documentation for descriptions of each of these
...
+  * When trigered, this type will retreive the alerts from the [Alerts API Endpoint](https://docs.opsgenie.com/docs/alert-api#list-alerts) producing on message per alert.
```
Comment (edreanernst) [inline README.md]:
```suggestion
  * When triggered, this action will retrieve a list of the alerts from the [Alerts API Endpoint](https://docs.opsgenie.com/docs/alert-api#list-alerts) producing one message per alert.
```

File: README.md [hunk 7]
```diff
@@ -129,4 +132,93 @@ Please refer to the Opsgenie API documentation for descriptions of each of these
...
+
+## Examples
+
+### Get current alerts
+```
+"msg": {
+    "action": "alerts"
```
Comment (edreanernst) [inline README.md]:
```suggestion
    "action": "list"
```

File: README.md [hunk 8]
```diff
@@ -129,4 +132,93 @@ Please refer to the Opsgenie API documentation for descriptions of each of these
...
+}
+```
+#### Example of an Msg produced per Alert
```
Comment (edreanernst) [inline README.md]:
```suggestion
#### Example of a message produced per alert
```

File: README.md [hunk 9]
```diff
@@ -129,4 +132,93 @@ Please refer to the Opsgenie API documentation for descriptions of each of these
...
+```
+"msg": {
+  "event_type": "OpsGenie Alert",
```
Comment (edreanernst) [inline README.md]:
```suggestion
  "event_type": "Opsgenie Alert",
```

File: README.md [hunk 10]
```diff
@@ -129,4 +132,93 @@ Please refer to the Opsgenie API documentation for descriptions of each of these
...
+  "topic": "OpsGenie Alert",
```
Comment (edreanernst) [inline README.md]:
```suggestion
  "topic": "Opsgenie Alert",
```

File: package.json [hunk 11]
```diff
@@ -1,12 +1,12 @@
 {
     "author": {
-        "name": "Edrean Ernst"
+        "name": "Edrean Ernst;Pedro Torrezão"
```
Comment (edreanernst) [inline package.json]: "author" field should include only one person, while "contributors" may be an array of persons
```suggestion
        "name": "Edrean Ernst"
```

File: package.json [hunk 12]
```diff
@@ -1,12 +1,12 @@
...
     },
```
Comment (edreanernst) [inline package.json:4]:
```suggestion
    },
    "contributors": [
        {
            "name": "Pedro Torrezão"
        }
    ]
```

File: package.json [hunk 13]
```diff
@@ -1,12 +1,12 @@
...
     "bugs": {
         "url": "https://github.com/edreanernst/node-red-contrib-opsgenie/issues"
     },
     "name": "node-red-contrib-opsgenie",
-    "version": "0.1.0",
+    "version": "0.1.1",
```
Comment (edreanernst) [inline package.json]: I will change the version later
```suggestion
    "version": "0.1.0",
```

File: src/opsgenie_alert_read.html [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+        category: 'Opsgenie',
+        color: '#2684ff',
+        defaults: {
+            parameters: {type:"opsgenie-config", required:true},
+            name: {value:""}
+        },
+        inputs:1,
+        outputs:1,
+        paletteLabel: "Opsgenie Alert Read",
+        icon: "font-awesome/fa-bell",
```
Comment (edreanernst) [inline src/opsgenie_alert_read.html]: Let's keep all the alert nodes the same in terms of icon
```suggestion
        icon: "alert.png",

```

File: src/opsgenie_alert_read.html [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+</script>
+
+
+<script type="text/x-red" data-template-name="OpsgenieAlertRead">
+    <div class="form-row">
+        <label for="node-input-parameters"><i class="fa fa-plug"></i> Parameters</label>
+        <input type="text" id="node-input-parameters">
+    </div>
+    <div class="form-row">
+        <label for="node-input-name"><i class="fa fa-tag"></i> Name</label>
```
Comment (edreanernst) [inline src/opsgenie_alert_read.html]:
```suggestion
        <label for="node-input-name"><i class="icon-tag"></i> Name</label>

```

File: src/opsgenie_alert_read.html [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+        <input type="text" id="node-input-name" placeholder="Name">
+    </div>
+</script>
+
+<script type="text/html" data-help-name="OpsgenieAlertRead">
+    <p>A node that retreives Opsgenie Alerts and turn them into messages (1 per alert)</p>
```
Comment (edreanernst) [inline src/opsgenie_alert_read.html]:
```suggestion
    <p>A node that retrieves Opsgenie alerts and turn them into messages (1 per alert)</p>

```

File: src/opsgenie_alert_read.js [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+
+        var og = new OG(node.params.apikey, node.params.version, node.params.apihost);
+
+        node.on('input', function (msg) {
+            
+            if (node.params == null || 
+                node.params.apikey == undefined || 
+                node.params.version == undefined || 
+                node.params.apihost == undefined) {
+                node.error("No configurations defined!");
```
Comment (edreanernst) [inline src/opsgenie_alert_read.js]:
```suggestion
                node.error("No configuration defined!");

```

File: src/opsgenie_alert_read.js [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+                return;
+            }
+
+            if (msg.action == undefined) {
+                node.error("No action defined!");
+                return;
+            }
+
+            switch (msg.action) {
+                case 'alerts':
```
Comment (edreanernst) [inline src/opsgenie_alert_read.js]:
```suggestion
                case 'list':

```

File: src/opsgenie_alert_read.js [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+                    try {
+                        og.getAlerts().then((aResult) => {
+                            if (aResult.success) {
+                                aResult.body.data.forEach(element => {
+                                    node.send({
+                                        event_type: "OpsGenie Alert",
```
Comment (edreanernst) [inline src/opsgenie_alert_read.js]:
```suggestion
                                        event_type: "Opsgenie Alert",

```

File: src/opsgenie_alert_read.js [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+                                        topic: "OpsGenie Alert",
```
Comment (edreanernst) [inline src/opsgenie_alert_read.js]:
```suggestion
                                        topic: "Opsgenie Alert",

```

File: src/opsgenie_alert_read.js [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+                                        payload: element.message,
+                                        data: element
+                                    });
+                                });
+                            } else {
+                                node.error("There was a problem reading an Opsgenie alerts: '" + aResult.error + "' - " + JSON.stringify(aResult.body));
```
Comment (edreanernst) [inline src/opsgenie_alert_read.js]:
```suggestion
                                node.error("There was a problem reading the Opsgenie alerts: '" + aResult.error + "' - " + JSON.stringify(aResult.body));

```

File: src/opsgenie_alert_read.js [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+                            }
+                        });
+
+                    } catch (error) {
+                        node.error("There was a problem reading an Opsgenie alerts: '" + error.message + "'");
```
Comment (edreanernst) [inline src/opsgenie_alert_read.js]:
```suggestion
                        node.error("There was a problem reading the Opsgenie alerts: '" + error.message + "'");

```

**Review by edreanernst — CHANGES_REQUESTED** [review]
Thanks for the changes!

I have made some suggested changes, which you can just accept if you're happy with it. Most are just typos and grammatical changes.

**Please also remove the node_modules folder from the PR.** It shouldn't be included in source control.

Please also add detailed documentation to the opsgenie_alert_read.html file. Please see opsgenie_alert_action.html for reference.

**Comment by ptorrezao** [comment]
Hi @edreanernst,
I did the changes that you requested. Sent it to review again.
Sorry for the grammar errors, the only excuse that I have is the fact that I'm not an Native Speaker (🤷‍♂️poor excuse, I know🤷‍♂️)

**Review by edreanernst — APPROVED** [review]

**Comment by edreanernst** [comment]
Thanks again @ptorrezao for the contribution! Sorry for the delay in processing the PR.

I have accepted and merged your contribution. You should see an official release soon.

Thanks again and any further contributions will be most welcome.

**Pull request merged**